File: module_book/src/main/java/com/ebook/book/mvvm/model/BookCommentsModel.java
package com.ebook.book.mvvm.model;

import android.app.Application;

import com.ebook.api.RetrofitManager;
import com.ebook.api.dto.RespDTO;
import com.ebook.api.entity.Comment;
import com.ebook.api.http.RxAdapter;
import com.ebook.api.service.CommentService;
import com.ebook.common.mvvm.model.BaseModel;

import java.util.List;

import io.reactivex.Observable;

public class BookCommentsModel extends BaseModel {
    private CommentService commentService;

    public BookCommentsModel(Application application) {
        super(application);
        commentService = RetrofitManager.getInstance().getCommentService();
    }

    /**
     * 添加评论
     */
    @SuppressWarnings("unchecked")
    public Observable<RespDTO<Comment>> addComment(Comment comment) {
        return commentService.addComment(RetrofitManager.getInstance().TOKEN, comment)
                .compose(RxAdapter.schedulersTransformer())
                .compose(RxAdapter.exceptionTransformer());
    }

    /**
     * 获得章节评论
     */
    @SuppressWarnings("unchecked")
    public Observable<RespDTO<List<Comment>>> getChapterComments(String chapterUrl) {
        return commentService.getChapterComments(RetrofitManager.getInstance().TOKEN, chapterUrl)
                .compose(RxAdapter.schedulersTransformer())
                .compose(RxAdapter.exceptionTransformer());
    }

    /**
     * 删除评论
     */
    @SuppressWarnings("unchecked")
    public Observable<RespDTO<Integer>> deleteComment(Long id) {
        return commentService.deleteComment(RetrofitManager.getInstance().TOKEN, id)
                .compose(RxAdapter.schedulersTransformer())
                .compose(RxAdapter.exceptionTransformer());
    }
}

File: module_book/src/main/java/com/ebook/book/mvvm/model/BookCommentsModel.kt
package com.ebook.book.mvvm.model

import android.app.Application
import com.ebook.api.entity.Comment
import com.ebook.api.service.comment.CommentDataSource
import com.ebook.api.utils.CoroutineAdapter
import com.xrn1997.common.dto.RespDTO
import com.xrn1997.common.manager.RetrofitManager
import com.xrn1997.common.mvvm.model.BaseModel
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class BookCommentsModel @Inject constructor(
    application: Application,
    private val dataSource: CommentDataSource
) : BaseModel(application) {
    /**
     * 添加评论
     */
    suspend fun addComment(comment: Comment): Result<RespDTO<Comment>> =
        CoroutineAdapter.safeApiCall {
            dataSource.addComment(RetrofitManager.TOKEN, comment)
        }


    /**
     * 获得章节评论
     */
    suspend fun getChapterComments(chapterUrl: String?): Result<RespDTO<List<Comment>>> =
        CoroutineAdapter.safeApiCall {
            dataSource.getChapterComments(RetrofitManager.TOKEN, chapterUrl)
        }

    /**
     * 删除评论
     */
    suspend fun deleteComment(id: Long): Result<RespDTO<Int>> =
        CoroutineAdapter.safeApiCall {
            dataSource.deleteComment(RetrofitManager.TOKEN, id)
    }

}
